packages/xforms-engine/src/lib/codecs/Geopoint/Geopoint.ts

```diff
@@ -0,0 +1,117 @@
+ * It can be expanded later to support units (e.g., degrees or meters),
+ * which would also serve as documentation to clarify what each value represents.
+ */
+abstract class SemanticValue<Semantic extends string, Value extends number | null> {
+	abstract readonly semantic: Semantic;
+
+	constructor(readonly value: Value) {}
+}
+
+class Latitude extends SemanticValue<'latitude', number> {
+	readonly semantic = 'latitude';
+}
+
+class Longitude extends SemanticValue<'longitude', number> {
+	readonly semantic = 'longitude';
+}
+
+class Altitude<Value extends number | null = number> extends SemanticValue<'altitude', Value> {
+	readonly semantic = 'altitude';
+}
+
+class Accuracy<Value extends number | null = number> extends SemanticValue<'accuracy', Value> {
+	readonly semantic = 'accuracy';
+}
+
+export interface GeopointValue {
+	readonly latitude: number;
+	readonly longitude: number;
+	readonly altitude: number | null;
+	readonly accuracy: number | null;
+}
+
+export type GeopointRuntimeValue = GeopointValue | null;
+
+interface GeopointInternalValue {
+	readonly latitude: Latitude;
+	readonly longitude: Longitude;
+	readonly altitude: Altitude<null> | Altitude<number>;
+	readonly accuracy: Accuracy<null> | Accuracy<number>;
+}
+
+type GeopointTuple =
+	| readonly [
+			latitude: Latitude,
+			longitude: Longitude,
+			altitude: Altitude<null> | Altitude<number>,
+			accuracy: Accuracy,
+	  ]
+	| readonly [latitude: Latitude, longitude: Longitude, altitude: Altitude]
+	| readonly [latitude: Latitude, longitude: Longitude];
+
+const DEGREES_MAX = {
+	latitude: 90,
+	longitude: 180,
+} as const;
+
+type CoordinateType = keyof typeof DEGREES_MAX;
+
+export class Geopoint {
+	private readonly internalValue: GeopointInternalValue;
+
+	constructor(coordinates: GeopointValue) {
+		const { latitude, longitude, altitude, accuracy } = coordinates;
+
+		this.internalValue = {
+			latitude: new Latitude(latitude),
+			longitude: new Longitude(longitude),
+			altitude: this.isValidNumber(altitude) ? new Altitude(altitude) : new Altitude(null),
+			accuracy: this.isValidNumber(accuracy) ? new Accuracy(accuracy) : new Accuracy(null),
+		};
+	}
+
+	getTuple(): GeopointTuple {
+		const { latitude, longitude, altitude, accuracy } = this.internalValue;
+
+		if (altitude.value == null && accuracy.value != null) {
+			return [latitude, longitude, altitude, accuracy];
+		}
+
+		if (altitude.value != null && accuracy.value != null) {
+			return [latitude, longitude, altitude, accuracy];
+		}
+
+		if (altitude.value != null) {
+			return [latitude, longitude, altitude];
+		}
+
+		return [latitude, longitude];
+	}
+
+	getRuntimeValue(): GeopointRuntimeValue {
+		const { latitude, longitude, altitude, accuracy } = this.internalValue;
+		const isLatitude = this.isValidDegrees('latitude', latitude.value);
+		const isLongitude = this.isValidDegrees('longitude', longitude.value);
+
+		if (!isLatitude || !isLongitude) {
+			return null;
+		}
```

Comment:
Just an observation, no action requested!

I hadn't thought about this case when I was pseudocoding out the tuple and semantic value concepts. Normally, I'd want to handle the invalid cases during (error) or prior to (return `null` from a factory) construction.

But I think this is the right call for now, **please don't change anything about this yet!** We'll need to revisit this as we incorporate the other codec use cases, and this leaves a good starting point to think about how to encode error conditions _into the value_ (presumably within its `internalValue`), which will be important as `Geopoint` is composed into `Geotrace` and `Geoshape` (and potentially `Line` distinct from `Geotrace`, for the XPath cases).